### ISSUE TITLE
The pkgconfig module does not check for `name_prefix`

### ISSUE THREAD

**Issue by arteymix** [mesonbuild/meson#889]
The generator simply does not check for the library prefix.

The case I deal with is when the package name starts with `lib`:

```python
libfoo_lib = library('libfoo', 'libfoo.c', name_prefix: '')

pkgconfig.generate(name: 'libfoo', library: libfoo_lib)
```

```pkgconfig
...
Libs: -L${libdir} -llibfoo # should be simply -lfoo
```

I'll probably submit a patch tomorrow for this problem on #862.

**Comment by nirbheek** [comment]
pkg-config won't be able to find your library if it doesn't start with `lib`, so editing `name_prefix` and using pkg-config are two things you should not mix.

**Comment by arteymix** [comment]
From gcc manpage:

> The only difference between using an -l option and specifying a file name is that -l surrounds library with lib and .a and searches several directories.

I guess we can specify an explicit filename if `name_prefix` is set.

**Comment by nirbheek** [comment]
Sadly, that doesn't work either. I tried. You need to specify the full path to the library. The `-L` option only applies to libraries specified with `-l`.

This is only with GCC and Clang. MSVC will happily search for library files in the `/LIBPATH` directories specified.

**Comment by arteymix** [comment]
I guess that for these kind of use cases, it's better to put a literal string, which is already covered in #862.

It would be nice to warn though if it cannot work in any cases.

**Comment by nirbheek** [comment]
Btw, why do you do this?

```
libfoo_lib = library('libfoo', 'libfoo.c', name_prefix: '')
```

Wouldn't this create the same filenames as:

```
libfoo_lib = library('foo', 'libfoo.c')
```

?

**Comment by arteymix** [comment]
Yeah, for C it does not matter, but for Vala the generated VAPI follow the target name. I'm wrapping libMemcached and the package name should be `libmemcached-glib` for consistency. 

See here: https://github.com/arteymix/libmemcached-glib

This is why I think we should at least handle the case where it has the `lib` prefix and warn for others.

**Comment by nirbheek** [comment]
I see, that makes sense. I'll update my PR.

**Comment by nirbheek** [comment]
Actually, maybe we should also allow you to set the basename for the generated vapi files? This seems like a common requirement for VAPI files looking at the list on my system.

**Comment by arteymix** [comment]
I like the idea for it being consistent with the target name.

Maybe we can do the same trick for `--gir` in #634 and check if `--vapi` is given in `vala_args`. I would do the same for `--header` as well. This would give a lot of flexibility while preserving sensible defaults.

**Comment by arteymix** [comment]
Let say we warn unconditionally for `name_prefix` and instead provide these facilities?

**Comment by nirbheek** [comment]
I've already fixed the PR to not warn if the filename starts with `lib` since that makes sense generally too, so the way you do things right now won't break.

But about vapi, gir, Hname, etc, I suppose that would be easier to implement than adding kwargs for each of those, but I'm not sure if it makes sense (now that I've looked into the whole thing a bit more). We don't do that anywhere else (except for backwards compat), and I think it sets a bad precedent.

The point of `*_args` is to pass compiler arguments that are not supported (or will not be supported) by Meson. Using that as a hack to pass options to Meson is very strange, IMO.

**Comment by arteymix** [comment]
I agree too, but that mean we have to introduce a bunch of additional kwargs for the GIR, header, etc...

Also, it's not really a hack, it's just about making Meson sensible to some flags that would otherwise be generated automatically. Think of it, the `vala_args` has already precedence over the implicit `--header` flag, so the following would break the target output:

```python
library('foo', 'foo.vala', vala_args: '--header=libfoo.h') # Meson assumes 'foo.h'
```

Also, targets have very generic arguments and the only language-specifics are passed via `<lang>_args`. If we start adding language specificities outside that scope, it will be never-ending.

**Comment by nirbheek** [comment]
The main reason why we have kwargs at all is because that's the only way to guarantee compatibility across compilers. That's not an issue with Vala, but it still sets a bad precedent and ties our API with the compiler's API.

`*_args` is how Meson gives you rope to hang yourself with. If you do stupid things with it, everything will break. The same is true for `c_args`, for instance. You can pass `-o` and your build breaks.

It may look like `*_args` is the recommended way to pass language-specific arguments, but we have have language-specific kwargs too: `cpp_pch`. The only reason why we don't have more is because our support for other languages is a WIP.

I don't think there's any significant cost to adding lots of kwargs, so I don't see a problem with adding more language-specific kwargs.

Anyway that's my reasoning. It's up to Jussi in the end.

**Comment by nirbheek** [comment]
There is another piece left here: `l.name` in this case would still be `libfoo`, so the pkg-config file would still have `-llibfoo` in it. I realized that later, and was working on a fix for that. Reopening till I have that fixed.

**Comment by arteymix** [comment]
They I guess we need to introduce some other arguments. I was reluctant at first, but if it's okay to introduce some of them, why not?

With `vala_vapi`, `vala_header` and `vala_gir` arguments, we should cover most if not all use cases.

**Comment by arteymix** [comment]
Oh, yeah, I forgot to mention the `-llib` issue. If we are not to warn for the `lib` prefix, it has to be stripped.

**Comment by nirbheek** [comment]
What is this `-llib` issue? Is that the same one that this bug was filed for? That should be fixed by #899.

**Comment by arteymix** [comment]
Yes, I meant that one. I was blocked at this point actually so thanks :+1:
